flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/AbstractCommitterHandler.java

```diff
@@ -19,18 +19,58 @@
 
 import org.apache.flink.util.function.SupplierWithException;
 
+import java.io.IOException;
 import java.util.ArrayDeque;
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.Deque;
 import java.util.List;
 
-abstract class AbstractCommitterHandler<InputT, OutputT>
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+abstract class AbstractCommitterHandler<InputT, OutputT, RecoverT>
         implements CommitterHandler<InputT, OutputT> {
 
     /** Record all the committables until commit. */
     private final Deque<InputT> committables = new ArrayDeque<>();
 
+    /** The committables that need to be committed again after recovering from a failover. */
+    private final List<RecoverT> recoveredCommittables = new ArrayList<>();
```

Comment:
Please note that for streaming we additionally also retry with every regular commit (as in your proposal). So, this approach in this PR is combining both ideas and is thus strictly superior in that regard to just retry with next commit. 
I'm assuming that you overlooked that both ideas are simultaneously used as your argument regarding OOM is also applying to the other approach (even if it's quite theoretical from my point of view) and even more pronounced there, since fewer retries would mean that the probability of accumulation increases.